sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/MetricName.java

```diff
@@ -17,63 +17,102 @@
  */
 package org.apache.beam.sdk.metrics;
 
-import static org.apache.beam.vendor.guava.v20_0.com.google.common.base.Preconditions.checkArgument;
+import static org.apache.beam.sdk.metrics.MetricUrns.USER_COUNTER_URN_PREFIX;
 
 import com.google.auto.value.AutoValue;
 import java.io.Serializable;
+import java.util.List;
+import javax.annotation.Nullable;
 import org.apache.beam.sdk.annotations.Experimental;
 import org.apache.beam.sdk.annotations.Experimental.Kind;
-import org.apache.beam.vendor.guava.v20_0.com.google.common.base.Strings;
+import org.apache.beam.vendor.guava.v20_0.com.google.common.base.Splitter;
 
 /**
- * The name of a metric consists of a {@link #getNamespace} and a {@link #getName}. The {@link
- * #getNamespace} allows grouping related metrics together and also prevents collisions between
- * multiple metrics with the same name.
+ * Wrapper for {@link org.apache.beam.model.pipeline.v1.MetricsApi.MonitoringInfo} URN.
+ *
+ * <p>"User" metrics (URN {@link MetricUrns#USER_COUNTER_URN_PREFIX}) are defined by a "namespace"
+ * and "name", and are the most commonly dealt with by user code, so structured constructors and
+ * accessors are provided in terms of those strings.
+ *
+ * <p>The {@link #namespace} allows grouping related metrics together and also prevents collisions
+ * between multiple metrics with the same name.
  */
 @Experimental(Kind.METRICS)
 @AutoValue
 public abstract class MetricName implements Serializable {
 
-  /** The namespace associated with this metric. */
-  public abstract String getNamespace();
+  public abstract String urn();
+
+  @Nullable private String name;
 
-  /**
-   * The namespace associated with this metric.
-   *
-   * @deprecated to be removed once Dataflow no longer requires this method.
-   */
-  @Deprecated
-  public String namespace() {
-    return getNamespace();
+  @Nullable private String namespace;
+
+  /** Parse the urn field into a name and namespace field. */
+  private boolean parseUrn(boolean raise) {
+    if (namespace != null && name != null) {
+      return true;
+    }
+    String urn = urn();
+    if (urn.startsWith(USER_COUNTER_URN_PREFIX)) {
+      urn = urn.substring(USER_COUNTER_URN_PREFIX.length());
+    } else {
+      if (raise) {
+        throw new IllegalStateException(
+            "Attempting to access namespace/name of a non-user metric: " + urn);
+      }
+      return false;
+    }
+    // If it is not a user counter, just use the first part of the URN, i.e. 'beam'
+    List<String> pieces = Splitter.on(':').splitToList(urn);
```

Comment:
We got duplication of this logic:
in this PR: MetricUrns.java L66
and in [DF runner](https://github.com/apache/beam/blob/fb425d5c4a7607d78f879e790ddebdc7bb5ea68a/runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/fn/control/UserMonitoringInfoToCounterUpdateTransformer.java#L106)

can be more.

Reply:
Thanks for flagging this!

I've changed this code (and [`MetricUrns.parseUrn`](https://github.com/apache/beam/pull/7915/files#diff-2a6c540e8bf410b089793ea6efadb291R61)) to match the semantics there ("namespace is everything up to the last `:`, with internal `:` deleted"), and [changed that site](https://github.com/apache/beam/pull/7915/files#diff-a2189a7c97dace5f44f04e16a71b6468R108) to call `MetricUrns.parseUrn`.

Combining this with MetricUrns.parseUrn would also be nice, but they're pretty different contexts and I don't see an easy way to atm.